server/src/main/java/com/linecorp/centraldogma/server/internal/mirror/credential/AbstractMirrorCredential.java

```diff
@@ -27,45 +28,83 @@
 
 import javax.annotation.Nullable;
 
+import com.fasterxml.jackson.annotation.JsonProperty;
 import com.google.common.base.MoreObjects;
 import com.google.common.base.MoreObjects.ToStringHelper;
 import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Iterables;
 
 import com.linecorp.centraldogma.server.mirror.MirrorCredential;
 
-abstract class AbstractMirrorCredential implements MirrorCredential {
+public abstract class AbstractMirrorCredential implements MirrorCredential {
+
+    /**
+     * The index should be updated by {@link #setIndex(int)} before use.
+     */
+    private int index = -1;
 
     @Nullable
     private final String id;
+    private final String type;
     private final Set<Pattern> hostnamePatterns;
     private final Set<String> hostnamePatternStrings;
+    private final boolean enabled;
 
-    AbstractMirrorCredential(@Nullable String id, @Nullable Iterable<Pattern> hostnamePatterns) {
+    AbstractMirrorCredential(@Nullable String id, String type, @Nullable Iterable<Pattern> hostnamePatterns,
+                             @Nullable Boolean enabled) {
         this.id = id;
+        // JsonTypeInfo is ignored when serializing collections.
+        // As a workaround, manually set the type hint to serialize.
+        this.type = requireNonNull(type, "type");
         this.hostnamePatterns = validateHostnamePatterns(hostnamePatterns);
         hostnamePatternStrings = this.hostnamePatterns.stream().map(Pattern::pattern)
                                                       .collect(Collectors.toSet());
+        this.enabled = firstNonNull(enabled, true);
     }
 
     private static Set<Pattern> validateHostnamePatterns(@Nullable Iterable<Pattern> hostnamePatterns) {
-        if (hostnamePatterns == null) {
+        if (hostnamePatterns == null || Iterables.isEmpty(hostnamePatterns)) {
+            return ImmutableSet.of();
+        }
+        if (Iterables.size(hostnamePatterns) == 1 && Iterables.getFirst(hostnamePatterns, null) == null) {
+            // Web browsers may send an empty array `[]` as a value of hostnamePatterns
+            // which is converted into `[null]`.
```

Comment:
Question) Is it possible for browsers to clean the input before sending such parameters? This feels unintuitive since the Web UI isn't the only code-path which uses this logic.

Reply:
It is possible. I can inject a custom logic to covert an empty array to null for `hostnamePatterns.